packages/liveblocks-core/src/lib/AsyncCache.ts

```diff
@@ -0,0 +1,345 @@
+import type { Callback, Observable, UnsubscribeCallback } from "./EventSource";
+import { makeEventSource } from "./EventSource";
+
+const DEDUPLICATION_INTERVAL = 2000;
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
+
+type AsyncFunction<T, A extends any[] = any[]> = (...args: A) => Promise<T>;
+
+type AsyncCacheOptions = {
+  deduplicationInterval?: number;
+};
+
+type AsyncCacheItemOptions = WithRequired<
+  AsyncCacheOptions,
+  "deduplicationInterval"
+>;
+
+type InvalidateOptions<TData = any> =
+  | { clearData?: false; setOptimisticData?: never }
+  | {
+      clearData?: never;
+      setOptimisticData: (data: TData | undefined) => TData | undefined;
+    };
+
+export type AsyncState<TData = any, TError = any> = {
+  isLoading: boolean;
+  data?: TData;
+  error?: TError;
+};
+
+type AsyncResolvedState<TData = any, TError = any> = AsyncState<
+  TData,
+  TError
+> & {
+  isLoading: false;
+};
+
+type AsyncCacheItemContext<TData, TError> = AsyncState<TData, TError> & {
+  isInvalid: boolean;
+  scheduledInvalidation?: InvalidateOptions<TData>;
+  rollbackOptimisticDataOnError?: boolean;
+  promise?: Promise<TData>;
+  lastInvokedAt?: number;
+  previousState: AsyncState<TData, TError>;
+  previousNonOptimisticData?: TData;
+};
+
+export type AsyncCacheItem<TData = any, TError = any> = Observable<
+  AsyncState<TData, TError>
+> & {
+  get(): Promise<AsyncResolvedState<TData, TError>>;
+  getState(): AsyncState<TData, TError>;
```

Comment:
Since React introduced the `useSyncExternalStore` API, they talk about `.getSnapshot()` for things like this. Here, it looks like `.getState()` is exactly that: a synchronous snapshot of the current state of this cache item. Shall we rename this method?

Reply:
It's precisely here for the `useSyncExternalStore`-based React implementation that's coming soon but since it can be used for other synchronous use-cases (e.g. advanced/custom revalidation logic) and the `isLoading`/`data`/`error` type is called `AsyncState` I thought it would make more sense to map to that instead of just the React use-case.